generators/generator-base-private.js

```diff
@@ -1203,59 +1204,41 @@ module.exports = class JHipsterBasePrivateGenerator extends Generator {
-    } else if (databaseType === MSSQL) {
-      if (protocol === 'r2dbc') {
-        dbcUrl = `${protocol}:mssql://${options.hostname}:1433/${options.databaseName}`;
-      } else {
-        dbcUrl = `${protocol}:sqlserver://${options.hostname}:1433;database=${options.databaseName}`;
-      }
-    } else if (databaseType === H2_DISK) {
-      if (!options.localDirectory) {
-        throw new Error(`'localDirectory' option should be provided for ${databaseType} databaseType`);
-      }
-      if (protocol === 'r2dbc') {
-        dbcUrl = `${protocol}:h2:file://${options.localDirectory}/${options.databaseName}`;
-      } else {
-        dbcUrl = `${protocol}:h2:file:${options.localDirectory}/${options.databaseName}`;
-      }
-    } else if (databaseType === H2_MEMORY) {
-      if (protocol === 'r2dbc') {
-        dbcUrl = `${protocol}:h2:mem:///${options.databaseName}${options.itests ? ':12344' : ''}`;
-      } else {
-        dbcUrl = `${protocol}:h2:mem:${options.databaseName}${options.itests ? ':12344' : ''}`;
-      }
+    let databaseDataForType = databaseData[databaseType];
+    if (databaseDataForType[protocol]) {
+      databaseDataForType = {
+        ...databaseDataForType,
+        ...databaseDataForType[protocol],
+      };
+    }
+    const { protocolSuffix = '', extraOptions = '', useDirectory = false } = databaseDataForType;
+    let { port = '' } = databaseDataForType;
+    if (useDirectory && !options.localDirectory) {
+      throw new Error(`'localDirectory' option should be provided for ${databaseType} databaseType`);
+    }
+    const databaseHasHost = options.hostname;
+    if (options.itests && H2_MEMORY === databaseType) {
+      port = ':12344';
+    }
+    let url = `${protocol}:${protocolSuffix}`;
+    if (options.localDirectory) {
+      url += `${options.localDirectory}/`;
     } else {
-      throw new Error(`${databaseType} databaseType is not supported`);
+      url += databaseHasHost ? options.hostname : options.databaseName;
+      url += port;
     }
-
-    if (!options.skipExtraOptions) {
-      dbcUrl += this.getDBCExtraOption(databaseType);
+    if (databaseHasHost || options.localDirectory) {
+      url += options.databaseName;
     }
-    return dbcUrl;
+    return `${url}${options.skipExtraOptions ? '' : extraOptions}`;
   }
 
   getDBCExtraOption(databaseType) {
-    let extraOptions = '';
-    if (databaseType === MYSQL) {
-      extraOptions =
-        '?useUnicode=true&characterEncoding=utf8&useSSL=false&useLegacyDatetimeCode=false&serverTimezone=UTC&createDatabaseIfNotExist=true';
-    } else if (databaseType === MARIADB) {
-      extraOptions = '?useLegacyDatetimeCode=false&serverTimezone=UTC';
-    } else if (databaseType === H2_DISK) {
-      extraOptions = ';DB_CLOSE_DELAY=-1';
-    } else if (databaseType === H2_MEMORY) {
-      extraOptions = ';DB_CLOSE_DELAY=-1;DB_CLOSE_ON_EXIT=FALSE';
-    }
+    const databaseDataForType = databaseData[databaseType];
```

Comment:
```suggestion
    let databaseDataForType = databaseData[databaseType];
    if (databaseDataForType[protocol]) {
      databaseDataForType = {
        ...databaseDataForType,
        ...databaseDataForType[protocol],
      };
    }
```

or create the function at `sql-constants.js`, I don't think it's released already, can be renamed to `sql-data.js`
```
function getDataForDatabase (databaseType, protocol) {
    let databaseDataForType = databaseData[databaseType];
    if (databaseDataForType[protocol]) {
      databaseDataForType = {
        ...databaseDataForType,
        ...databaseDataForType[protocol],
      };
    }
}

```

Reply:
Constants are constants ^^. This methods is used in testcontainers testcustomizer, so should be made available to the templates.
I plan to move all these methods to their dedicated js and use generator-base and base-private as a facade.